quickwit/quickwit-control-plane/src/ingest/ingest_controller.rs

```diff
@@ -0,0 +1,1348 @@
+        info!("syncing internal state with metastore");
+        let now = Instant::now();
+
+        self.index_table.clear();
+        self.shard_table.clear();
+
+        let indexes = ctx
+            .protect_future(self.metastore.list_indexes_metadatas(ListIndexesQuery::All))
+            .await?;
+
+        self.index_table.reserve(indexes.len());
+
+        let num_indexes = indexes.len();
+        let mut num_sources = 0;
+        let mut num_shards = 0;
+
+        let mut subrequests = Vec::with_capacity(indexes.len());
+
+        for index in indexes {
+            for source_id in index.sources.into_keys() {
+                if source_id != INGEST_SOURCE_ID {
+                    continue;
+                }
+                let request = metastore::ListShardsSubrequest {
+                    index_uid: index.index_uid.clone().into(),
+                    source_id,
+                    shard_state: Some(ShardState::Open as i32),
+                };
+                num_sources += 1;
+                subrequests.push(request);
+            }
+            self.index_table
+                .insert(index.index_config.index_id, index.index_uid);
+        }
+        if !subrequests.is_empty() {
+            let list_shards_request = metastore::ListShardsRequest { subrequests };
+            let list_shard_response = ctx
+                .protect_future(self.metastore.list_shards(list_shards_request))
+                .await?;
+
+            self.shard_table
+                .table_entries
+                .reserve(list_shard_response.subresponses.len());
+
+            for list_shards_subresponse in list_shard_response.subresponses {
+                num_shards += list_shards_subresponse.shards.len();
+
+                let key = (
+                    list_shards_subresponse.index_uid.into(),
+                    list_shards_subresponse.source_id,
+                );
+                let shard_entries: HashSet<ShardEntry> = list_shards_subresponse
+                    .shards
+                    .into_iter()
+                    .map(ShardEntry)
+                    .collect();
+                let table_entry = ShardTableEntry {
+                    shard_entries,
+                    next_shard_id: list_shards_subresponse.next_shard_id,
+                };
+                self.shard_table.table_entries.insert(key, table_entry);
+            }
+        }
+        info!(
+            "synced internal state with metastore in {} seconds ({} indexes, {} sources, {} \
+             shards)",
+            now.elapsed().as_secs(),
+            num_indexes,
+            num_sources,
+            num_shards,
+        );
+        Ok(())
+    }
+
+    /// Pings an ingester to determine whether it is available for hosting a shard. If a follower ID
+    /// is provided, the leader candidate is in charge of pinging the follower candidate as
+    /// well.
+    async fn ping_leader(
```

Comment:
I would rename this 

`ping_leader_and_follower`
or 
`ping_follower_through_leader`
or whatever you see fit.